### PR TITLE
zkvm/blockchain: implement BlockchainState::apply_txlog

### PR DESCRIPTION
**Pull request by vickiniu**
Changes UTXO representation to match contract ID and implements `apply_txlog`, removing inputs and adding outputs to the UTXO set.

### CODE REVIEW

File: zkvm/src/contract.rs [hunk 1]
```diff
@@ -118,6 +119,11 @@ impl ContractID {
         &self.0
     }
 
+    /// Returns the contractID as a UTXO
+    pub fn as_utxo(&self) -> UTXO {
+        UTXO(self.0)
```
Comment (bobg) [inline zkvm/src/contract.rs]: In ZkVM, utxo and contractID are synonyms. That doesn't mean we should have two distinct types! Can you replace uses of the UTXO type with the ContractID type?

File: zkvm/src/blockchain/state.rs [hunk 2]
```diff
@@ -45,19 +45,83 @@ impl BlockchainState {
     fn apply_txlog(&mut self, txlog: &TxLog) -> Result<(), VMError> {
         for entry in txlog.iter() {
             match entry {
-                Entry::Input(contractID) => {
-                    // xxx check contractID is in self.utxos
-                    // xxx remove contractID
+                // Remove input from UTXO set
+                Entry::Input(input) => {
+                    let utxo = input.as_utxo();
+                    if self.utxos.contains(&utxo) {
```
Comment (bobg) [inline zkvm/src/blockchain/state.rs]: Is there a more Rust-idiomatic way to do this whole if/else block? Something like:

`self.utxos.remove_or_die(&utxo)?`

The error returned by `remove_or_die` would have to be converted to a `VMError::InvalidInput`. Is there an idiom for that? If not, maybe this code is good the way it is.

File: zkvm/src/blockchain/state.rs [hunk 3]
```diff
@@ -45,19 +45,83 @@ impl BlockchainState {
...
+                Entry::Output(output0.clone()),
+                Entry::Output(output1.clone()),
+            ])
+            .unwrap();
+        state
+            .apply_txlog(&vec![Entry::Input(output0.id())])
+            .unwrap();
+
+        // Check that output0 was consumed
+        assert_eq!(state.utxos.contains(&output0.id().as_utxo()), false);
```
Comment (bobg) [inline zkvm/src/blockchain/state.rs]: Is there a reason to prefer this over `assert!(!state.utxos.contains(...))`?

File: zkvm/src/contract.rs (hunk 1 above)
Comment (vickiniu) [reply zkvm/src/contract.rs]: Makes sense! That's even simpler :)

File: zkvm/src/blockchain/state.rs (hunk 2 above)
Comment (vickiniu) [reply zkvm/src/blockchain/state.rs]: Good question! There might very well be such a thing

File: zkvm/src/blockchain/state.rs (hunk 3 above)
Comment (vickiniu) [reply zkvm/src/blockchain/state.rs]: Not really! I'm happy to go with that

Comment (bobg) [reply zkvm/src/blockchain/state.rs]: `!(!`

File: zkvm/src/blockchain/state.rs [hunk 4]
```diff
@@ -47,17 +47,14 @@ impl BlockchainState {
             match entry {
                 // Remove input from UTXO set
                 Entry::Input(input) => {
-                    let utxo = input.as_utxo();
-                    if self.utxos.contains(&utxo) {
-                        self.utxos.remove(&utxo);
-                    } else {
+                    if !self.utxos.remove(&input) {
```
[the remark this answers is not in the document]
Comment (vickiniu) [reply zkvm/src/blockchain/state.rs]: I think for `HashSet`, `remove` returns a `bool`! https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.remove

**Pull request merged**